Migrate/group set selector

returnMarcco opened the pull request:
# Description

- This PR contains changes for migration the `group-set-selector` component from Coffeescript/AngularJS to Typescript/Angular.

Fixes # (issue)
- https://planner.cloud.microsoft/webui/v1/plan/njykIFLDn0iAY1at7tACfcgADgBS/view/grid/task/d6Zm3rI4q0e4t5wvnIlrkcgAHwst?tid=d02378ec-1688-46d5-8540-1c28b5f470f6

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
     - **Important:** As this Angular component emits events using Angular syntax, the parent component's declaration of this component does not currently bind to this emitted event properly. A workaround would have been to declare this component in the parent using AngularJS syntax, but I chose to use Angular syntax as the parent component can be upgraded to Angular next with less confusion. I will leave a note regarding this issue in the Planner ticket for the parent component.
- [x] This change requires a documentation update (README migration todo list)
- [x] Frontend migration

# How Has This Been Tested?

- Logging in as an admin
- Navigation to the `Unit` settings
- Selecting a unit
- Navigating to `Groups`
- Testing `Unit Group` dropdown functionality and event handling by stepping through the code in devtools
- Confirming that the parent component is not properly handling the emitted event when selecting a group set from the dropdown. 

- Note: This dropdown may not be visible by default. To make it visible for testing purposes, find its selector declaration and change the `ngIf` value to true.

# Screenshots

Before
<img width="980" height="232" alt="group_set_selector_before" src="https://github.com/user-attachments/assets/e8de9cb4-72e5-4d9d-896a-e2baf349c6c6" />

After
<img width="766" height="783" alt="Screenshot 2025-08-03 200834" src="https://github.com/user-attachments/assets/76f5baa9-00f7-4208-9be4-e158fa6c477a" />

## Testing Checklist:

- [x] Tested in latest Chrome
- [ ] Tested in latest Safari
- [ ] Tested in latest Firefox

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have requested a review from @macite and @jakerenzella on the Pull Request

returnMarcco committed:
chore: migrate group-set-selector (WIP)

- define and declare new component files

- import new component into app

- downgrade component

returnMarcco committed:
chore: migrate group-set-selector

- convert AngularJS template syntax to Angular syntax
- finish implementing `group-set-selector.component.ts`
- remove redundant Coffeescript file

returnMarcco committed:
chore: migrate group-set-selector

- finish defining Typescript logic
- update parent component to use Angular syntax for child component declaration

returnMarcco committed:
chore: cleanup and documentation

- remove redundant Coffeescript file
- adjust frontend migration todo list

returnMarcco committed:
chore: remove redundant file

- remove redundant Coffeescript file

jtalev commented on src/app/groups/group-set-selector/group-set-selector.component.html:
```diff
@@ -0,0 +1,8 @@
+  class="form-control"
+  [(ngModel)]="selectedGroupSet"
+  (change)="selectGroupSet($event.target.value)">
+  <option *ngFor="let gs of unit?.groupSets" [value]="gs">
```
using (ngModelChange) rather than (change) and only sending $event rather than $event.target.value, as well as [ngValue] rather than [value] seems to work, without doing this i could not get functionality to work so even if you want to stick with the angular syntax in the parent i think it is worth making this change

jtalev commented on src/app/groups/group-selector/group-selector.tpl.html:
```diff
@@ -3,10 +3,10 @@ <h4 class="panel-title">
+      *ngIf="showGroupSetSelector"
+      [unit]="unit"
+      [(selectedGroupSet)]="selectedGroupSet"
+      (onChange)="selectGroupSet($event)"
```
use [selected-group-set] and [on-change] to bind attributes in the parent

<img width="811" height="488" alt="angular" src="https://github.com/user-attachments/assets/5523255e-6fc6-43c3-b485-4c0305a414bb" />

jtalev commented:
Hey Jason, I've made a couple of comments on some of the code that could be reviewed. Just a question, I'm a bit confused about the reasoning to not use the angularJs syntax in the parent as this leaves the component in a broken state? otherwise the component displays fine, just that the functionality is currently broken. Also i think the old .tpl.html file is still in the codebase and can be deleted.

returnMarcco commented:
> Hey Jason, I've made a couple of comments on some of the code that could be reviewed. Just a question, I'm a bit confused about the reasoning to not use the angularJs syntax in the parent as this leaves the component in a broken state? otherwise the component displays fine, just that the functionality is currently broken. Also i think the old .tpl.html file is still in the codebase and can be deleted.

Hey Josh.  I went with the Angular syntax route to adhere to the pattern of using that syntax in Angular child components nested in an AngularJS parent, and to make the parent component upgrade a bit easier such that it would not need refactoring for the child component's attribute declarations.

 As the functionality of this component seems to be mimicked by clicking on each group set row, and also because it seems to be hidden by default, I reasoned that it could lay dormant until the parent component was upgraded.

On second thoughts, I agree that it should function as is and shouldn't be introduced in a broken state. I'll resolve your feedback and push my changes before re-requesting a review.

Cheers

returnMarcco committed:
fix: attribute and event binding

- fix attribute and event binding syntax
- remove redundant template file

returnMarcco commented:
Hi @jtalev ,

Thanks for your feedback. I've implemented the suggested changes and tested on my end with success.

Cheers

jtalev reviewed (approved):
Hey Jason, I've pulled your changes and everything looks good, great work!

disururathnayake commented:
I managed to pull your request and build it without any errors. Here is my feedback:

1. Successfully unlinked and removed the older instances of the component from doubtfire-angularjs.module.ts and downgrade the component.
2. Has successfully deleted old coffee file and new instance of the components has been created.
3. Successfully imported the new component into the doubtfire-angular.module.ts.
4. Successfully unlinked  the older instances of the component from src/app/groups/groups.coffee.
5. I followed the steps you provided to test the component and the component is working as expected without any errors.

Great job on the migration. Well done.

returnMarcco commented:
> I managed to pull your request and build it without any errors. Here is my feedback:
> 
> 1. Successfully unlinked and removed the older instances of the component from doubtfire-angularjs.module.ts and downgrade the component.
> 2. Has successfully deleted old coffee file and new instance of the components has been created.
> 3. Successfully imported the new component into the doubtfire-angular.module.ts.
> 4. Successfully unlinked  the older instances of the component from src/app/groups/groups.coffee.
> 5. I followed the steps you provided to test the component and the component is working as expected without any errors.
> 
> Great job on the migration. Well done.

Thanks for the review

lachlan-robinson reviewed (changes requested):
Hey @returnMarcco

- [x] Migration looks clean and successfully addresses all of the functionality of the directive.
- [x] Event emission and error handling are preserved.
- [x] Good use of Angular features such as Input/Output for handling data flow.
- [ ] Potential gap in binding following migration.
The old AngularJS used two-way binding via ng-model, thus, keeping the parent selectedGroupSet in sync automatically. Currently, without explicit handling of the emitted change by the parent component, the child state can be altered without the parent state updating accordingly.

returnMarcco commented:
> Hey @returnMarcco
> 
> * [x]  Migration looks clean and successfully addresses all of the functionality of the directive.
> * [x]  Event emission and error handling are preserved.
> * [x]  Good use of Angular features such as Input/Output for handling data flow.
> * [ ]  Potential gap in binding following migration.
>   The old AngularJS used two-way binding via ng-model, thus, keeping the parent selectedGroupSet in sync automatically. Currently, without explicit handling of the emitted change by the parent component, the child state can be altered without the parent state updating accordingly.

Hi @lachlan-robinson,

The child component (group-set-selector) is using `ngModel` for two-way binding of the `selectedGroupSet` property. When a value is selected in the dropdown, `onModelChange` is called which updates the child component's `selectedGroupSet` property, and then emits an event, and the parent updates due to it's `[(selected-group-set)]` binding. 

As far as I can tell, the parent won't need explicit event-handling logic for the emitted event. Can you please provide me an example incase I'm missing something.

Cheers

returnMarcco committed:
refactor: two-way binding

- remove redundant directive attributes

- rename EventEmitter var name in child component

returnMarcco committed:
Merge branch '9.x' into migrate/group-set-selector

returnMarcco committed:
Update README.md

returnMarcco committed:
Update README.md

BrianDangDev commented on src/app/groups/group-set-selector/group-set-selector.component.html:
a suggestion is to add the disabled condition if unit.groupSets is empty or undefined for better UX in my opinion.  otherwise eveything looks good

returnMarcco replied:
Hi Brian. That's a good suggestion. The parent component (`group-selector`) actually hides this component if the length of the `groupSet` collection is 0. This was the behavior in the original implementation.

BrianDangDev reviewed (commented):
Everything looks good, @returnMarcco share your opinions on my comment in the group-set-selector.component.html file

returnMarcco commented:
Raising an upstream PR now

returnMarcco committed:
refactor: replace Bootstrap components with Angular Material

- replace Bootstrap components with Angular Material components & directives

- refactor to use modern Angular structural directives
